File: examples/cpp/chunk_operator.cpp
#include <cstddef>
#include <cstdint>
#include <vector>

#include <tsl.hpp>

struct chunk_sum_op {
    const std::int32_t* base = nullptr;
    std::int64_t total = 0;
    std::size_t visited = 0;
    bool metadata_ok = true;

    template <class Vec>
    void operator()(
        const std::int32_t* ptr,
        std::size_t offset,
        std::size_t count) {
        std::size_t expected_count = 0;
        if constexpr (Vec::has_static_lane_count_v) {
            expected_count = Vec::lane_count_v;
        } else {
            expected_count = Vec::lane_count();
        }
        if ((ptr != base + offset) || (count != expected_count)) {
            metadata_ok = false;
        }
        const auto values = tsl::load<Vec, false>(ptr);
        total += static_cast<std::int64_t>(tsl::hadd<Vec>(values));
        visited += count;
    }
};

void fill_input(std::vector<std::int32_t>& input) {
    for (std::size_t i = 0; i < input.size(); ++i) {
        input[i] = static_cast<std::int32_t>(static_cast<int>((i * 29) % 73) - 36);
    }
}

std::int64_t expected_sum(const std::vector<std::int32_t>& input) {
    std::int64_t total = 0;
    for (const auto value : input) {
        total += value;
    }
    return total;
}

template <class Parallelism>
bool run_policy_case(const std::vector<std::int32_t>& input, std::int64_t expected) {
    chunk_sum_op op{input.data()};
    tsl::algo::for_each_chunk<Parallelism>(
        op,
        input.data(),
        input.size());
    return op.metadata_ok && (op.visited == input.size()) && (op.total == expected);
}

int main() {
    constexpr std::size_t count = 1003;
    std::vector<std::int32_t> input(count);
    fill_input(input);
    const auto expected = expected_sum(input);

    if (!run_policy_case<tsl::dataparallel::native>(input, expected)) {
        return 1;
    }
    if (!run_policy_case<tsl::dataparallel::fixed<1>>(input, expected)) {
        return 2;
    }
    if (!run_policy_case<tsl::dataparallel::generic<4>>(input, expected)) {
        return 3;
    }
    if (!run_policy_case<tsl::dataparallel::generic<16>>(input, expected)) {
        return 4;
    }
    return 0;
}

File: examples/rust/src/bin/chunk_operator.rs
use tsl::tsl_core::StaticSimdVector;
use tsl::profile;

struct ChunkSum {
    base: *const i32,
    total: i64,
    visited: usize,
    metadata_ok: bool,
}

impl ChunkSum {
    fn new(input: &[i32]) -> Self {
        Self {
            base: input.as_ptr(),
            total: 0,
            visited: 0,
            metadata_ok: true,
        }
    }

    fn is_valid(&self, expected_total: i64, expected_count: usize) -> bool {
        self.metadata_ok && self.total == expected_total && self.visited == expected_count
    }
}

impl<V> profile::algo::ChunkKernel<V> for ChunkSum
where
    V: StaticSimdVector<BaseType = i32>
        + profile::detail::primitives::LoadImpl<false>
        + profile::detail::primitives::HaddImpl,
{
    unsafe fn apply(&mut self, ptr: *const V::BaseType, offset: usize, count: usize) {
        let expected_ptr = unsafe { self.base.add(offset) };
        if ptr != expected_ptr || count != V::ELEMENT_COUNT {
            self.metadata_ok = false;
        }

        let values = unsafe { profile::load::<V, false>(ptr) };
        self.total += i64::from(profile::hadd::<V>(values));
        self.visited += count;
    }
}

fn fill_input(input: &mut [i32]) {
    for (i, value) in input.iter_mut().enumerate() {
        *value = ((i * 29) % 73) as i32 - 36;
    }
}

fn expected_sum(input: &[i32]) -> i64 {
    input.iter().map(|value| i64::from(*value)).sum()
}

fn main() {
    let mut input = vec![0i32; 1003];
    fill_input(&mut input);
    let expected = expected_sum(&input);

    let mut native = ChunkSum::new(&input);
    profile::algo::for_each_chunk(tsl::dataparallel::native(), &mut native, &input);
    assert!(native.is_valid(expected, input.len()));

    let mut fixed = ChunkSum::new(&input);
    profile::algo::for_each_chunk(tsl::dataparallel::fixed::<1>(), &mut fixed, &input);
    assert!(fixed.is_valid(expected, input.len()));

    let mut generic4 = ChunkSum::new(&input);
    profile::algo::for_each_chunk(
        tsl::dataparallel::generic::<4>(),
        &mut generic4,
        &input,
    );
    assert!(generic4.is_valid(expected, input.len()));

    let mut generic16 = ChunkSum::new(&input);
    profile::algo::for_each_chunk(
        tsl::dataparallel::generic::<16>(),
        &mut generic16,
        &input,
    );
    assert!(generic16.is_valid(expected, input.len()));
}
